### ISSUE TITLE
Add alerts on PACER dockets

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#835]
The trickiest thing here will probably be the UI and performance side of it, but I think we can probably swing it pretty quickly.

**Comment by mlissner** [comment]
OK, first thoughts:

1. We already have a processing pipeline that tracks new docket entries. I think we can tie these alerts to that.

2. Do we need the rate of these to be adjustable? I don't think so. I think we just send them immediately or as close to immediately as possible. 

3. If we get 100,000 new items per day, each one will require that we look up alerts for it, that's 100,000 more queries per day. That's not great, but the alert table will be small. I think we can live with it.

4. We'll want to batch these somehow. For example, if we get 10 docket entries for a case in five minutes, we should only send one alert. This could happen, for example, when a RECAP user gives us a docket we only had part of previously. 

5. I assume we only send alerts when there's *new* content for a docket. I.e., if we have docket entry 3 and then docket entry 2 comes in, we don't care — it's old stuff. 

6. I think we can do a button for one-click alerts on every docket page. You click it, you're signed up. If you're logged out, you're asked to sign in.

What else? There are probably a lot of other concerns here.

**Comment by mlissner** [comment]
OK, some thoughts after a day of ~stewing~ consideration:

1. The hard part about batching is knowing if you're in the middle of aggregating a batch. For example consider this series of events, where we assume we only send an email for a docket at most once/five minutes:

        12:01pm: New docket entry comes in. No alert should be sent right now, but since nothing has come in for the last five minutes, we should start a new batch of events to bundle into an email.
        12:05pm: New docket entry comes in. No alert sent, but batch mode detected and item added to batch.
        12:06pm: Five minutes since batch started, Email gets sent with two new items.
        12:10pm: New item comes in. No alert sent, but new batch is begun again.

    How do we know that at 12:05 we've already got a batch going? How do we make sure that if the server crashes at 12:04, we don't think we're in batch mode forever and never send another alert? 

    I think the answer to most of these questions is to leverage Redis's expiration timers. Thus:

     - At 12:01, we create a new celery task that, in exactly five minutes, will process everything after that moment. We also set a key in redis that has an expiration time of five minutes (I think the name of this key can just be `alert:docket:$docket_id` or similar). This key indicates that there is a batch in progress.
     - At 12:05, we get a new docket entry. We check for the key in redis, find it, and do not send an alert.
     - At 12:06, the celery alert triggers. It pulls everything from the DB that changed in the last five minutes and then sends an email with that information.
     - At 12:10, a new item comes in. Again, we check for the key in redis. It's missing, so we start a new batch.

    Question: At 12:06 when we send the alert, do we query the whole database, or do we keep a cache of new information we intend to send? Latter is (slightly) more complex, but also vastly more performant.

    I think this will solve the major architectural questions for this feature.

1. New question: Do we send alerts for new attachments? I think not, if we can avoid it. I think we should only care about new **docket entries**. I think the attachments are assumed. Otherwise, we'd get the docket entry, send an alert, and then send an alert a while later to say, "Oh, hey, there's a new attachment!" Well, not really. That attachment was from a while ago.

1. This implies that we only send based on docket entries, not recap documents. That makes things simpler. We have a `BigInteger` field being used for the entry numbers. It's not universally true that things get filed in order, but I think we can work with that and consider it good enough. The alternative is to use the `date_filed` field — worth considering too. That's probably a better approach, actually. More robust against crazy court stuff.

**Comment by mlissner** [comment]
>  But I do think [the batching interval] needs to be adjustable internally.

Reasonable, and easy. Good call.

> people can be annoyed if they're seeing something on social media or email from colleagues or even news stories and still haven't gotten their alerts yet.

I think we will just have to handle that via the help page for alerts, where we can explain why there might be a delay.

> What does small mean?

Well, for now, we're in the thousands, which would be no problem at all if we've got it indexed properly. I think we can cross this bridge later if we have to. It'll be a good problem to have.

> Well. You probably want the first one to go out and then the remaining 9 later.
Maybe with a notion that "further alerts within the next N minutes will be aggregated and sent later."
Although you could imagine an alert on the first, and alert on the 2nd with the aggregation note, and then the tenth after some point.

I'd rather keep this simple rather than make people think about it or even know it exists. If there's just a two minute delay, that'll catch the worst of these issues, and people probably won't notice it.

> Does the clock reset when a new item arrives? So receipt of a new item every 60 seconds for an hour will delay the alert indefinitely? What about an MDL case like cand's 3:16-md-02741-VC where maybe somebody decides to file one document for each plaintiff and there are several hundred?

I think this will be fine with the current design I have in mind. You'd get an alert every two minutes.

> This assumption does not seem to be justified. Especially if entries come in out of order.

I think this is enough of a corner case that I'm not sure it's worth dealing with. I'll see if there's a good way to do it, but I think this would cause more false alarms than useful signal.

> With batched alerts, do you do one email message per case? Multiple cases per email?

One email per case with as many batched docket entries as makes sense. Never more than one case per email — agreed. We actually do multiple alerts per email for our search alerts, but it makes more sense for daily kinds of alerts than it does for triggered ones like this.

> Let's talk about server crashes separately. 

These also sub for code crashes, which are much more likely. If the code crashes for some reason, it'll be nice if running it again just works without having to do cleanup. Redis will make that possible.

> "I'm waiting and waiting and waiting for somebody to upload the exhibits to X so I don't have to buy them for $3. I want to know when it happens."

Note that this doesn't support that. It only notifies when there are new docket entries. If you want to know when there's a new document, I think search alerts will be the way to do that. I also think this is enough of a corner case that we should just document it, and call that good enough. If people want it, they'll ask and we'll have to reconsider.

**Comment by mlissner** [comment]
Meant to post one more thing: I think we're mostly on the same page. Batching is the most controversial thing, but I think a two minute delay to accomplish batching isn't horrible for most use cases. I bet most people won't notice it, in fact.

Of to the code mines. Let's see how far I can move this particular rock tonight.

**Comment by johnhawkinson** [comment]
> Note that this doesn't support that. It only notifies when there are new docket entries. If you want to know when there's a new document, I think search alerts will be the way to do that. I also think this is enough of a corner case that we should just document it, and call that good enough. If people want it, they'll ask and we'll have to reconsider.

really? Let's talk about the most common cases.

I would think that "waiting for a free document to become available" is one of the most common. But case-based analysis is important.

Another is, "I want to know when there's a doc so I can buy it," of course.

I would especially like to hear your response to:

> Also, we should acknowledge that in many cases alert frequency is triggered by RSS feed update frequency, which ranges from 15 minutes for 2 hours (right?). But people can be annoyed if they're seeing something on social media or email from colleagues or even news stories and still haven't gotten their alerts yet.

**Comment by mlissner** [comment]
I think the most important use case is to keep up with a case — to learn when it has new filings. For that, you need to know when there are new docket entries. I don't see most people waiting around for free documents. If people are like me, they'd notice something new on a case, realize it's missing, and go buy it. If you're interested enough to follow a case, you're probably interested enough to buy documents. 

I don't see people interested in a case twiddling their thumbs waiting for somebody else to upload something. 

> Also, we should acknowledge that in many cases alert frequency is triggered by RSS feed update frequency, which ranges from 15 minutes for 2 hours (right?).

I don't know the actual frequencies, but there's not a whole lot we can do here. I think the best we can do is:

1. Put that in the documentation.
1. Use a reasonable buffer size for sending emails so we don't bombard people.
1. Eventually start crawling the iQuery pages for popular cases.

In other words, set expectations, strike a UX balance, and do better when we can.

**Comment by johnhawkinson** [comment]
> I don't know the actual frequencies, but there's not a whole lot we can do here. I think the best we can do is:

But the aggregation/buffering code seems like it will negatively interact if it delays alerts when the RSS feed interval is much greater than the aggregation buffer time span. That is, waiting for something that will almost never happen.

Maybe you're saying it will not get triggered in that case? I'm not confident of htat fomr your description but maybe it's so?

**Comment by mlissner** [comment]
What would happen is, we'd:

1. Crawl the latest RSS
1. We'd identify three new docket entries for a docket
1. We'd start a batch email that'd go out in two minutes
1. If anything comes in in the next two minutes, it gets added to the batch
1. We crawl the RSS again, identify another two docket entries. They get added to the batch.
1. After two minutes from step 2, above, the email goes out
1. We crawl the RSS again. If we get anything new, we start a new two minute batch.

So, yeah, it could add a two minute delay over scraping the RSS yourself, but I think that's an OK compromise to save yourself from extra emails. 

Maybe a better trick would be to make the batch adjustable according to the use case. If RSS comes in, we wait 20s. If we get a docket uploaded, we wait two minutes.

**Comment by johnhawkinson** [comment]
> So, yeah, it could add a two minute delay over scraping the RSS yourself, but I think that's an OK compromise to save yourself from extra emails.

That just doesn't make sense. If a group comes in via RSS, they should not be delayed, because the RSS will already batch and delay them.

**Comment by mlissner** [comment]
Yeah, I think the only delay I envisioned there (as of my last message; story developing) was to make sure that if the first item in the RSS didn't trigger the alert, and then have the second item in the RSS trigger it again. For that, maybe we need 10s to process an entire RSS document. But if we can batch process these, which I think we can, you're right, no delay needed for RSS.

**Comment by johnhawkinson** [comment]
OK. If that's where we are probably a semaphore is better than a time delay.

**Comment by mlissner** [comment]
Yeah, that's kinda where the code is headed.

**Comment by mlissner** [comment]
Marking this done. Some details to fine tune, but it's basically there.
